src/core/common/bitflags.hpp

```diff
@@ -0,0 +1,187 @@
+
+    /**
+     * This method indicates whether all of the given bits are set in the bit flags.
+     *
+     * @param[in] aVals  The bits.
+     *
+     * @retval TRUE   If the given bits are all set.
+     * @retval FALSE  If the given bits are not all set.
+     *
+     */
+    template <typename... T0> bool HasAll(T0... aVals)
+    {
+        BaseType expectedBits = BuildBits(GetBit(aVals)...);
+
+        return (mBits & expectedBits) == expectedBits;
+    }
+
+    /**
+     * This method sets the given bits.
+     *
+     * @param[in] aVals  The bits.
+     *
+     */
+    template <typename... T0> void Set(T0... aVals) { mBits |= BuildBits(GetBit(aVals)...); }
+
+    /**
+     * This method unsets the given bits.
+     *
+     * @param[in] aVals  The bits.
+     *
+     */
+    template <typename... T0> void Unset(T0... aVals) { mBits &= ~BuildBits(GetBit(aVals)...); }
+
+    /**
+     * This method indicates whether any of the given bit is set in the bit flags.
+     *
+     * @param[in] aVals  The bits.
+     *
+     * @retval TRUE   If any of the given bits is all set.
+     * @retval FALSE  If none of given bits are all set.
+     *
+     */
+    template <typename... T0> bool HasAny(T0... aVals) { return (mBits & BuildBits(GetBit(aVals)...)) != 0; }
+
+    /**
+     * This method indicates whether all of and only the given flags are set.
+     *
+     * @param[in] aVals  The bits.
+     *
+     * @retval TRUE   If the given bits are all set.
+     * @retval FALSE  If the given bits are not all set.
+     *
+     */
+    template <typename... T0> bool HasExactly(T0... aVals) { return mBits == BuildBits(GetBit(aVals)...); }
+
+    /**
+     * This method indicates whether none of the given flags are set.
+     *
+     * @param[in] aVals  The bits.
+     *
+     * @retval TRUE   If the given bits are all set.
+     * @retval FALSE  If the given bits are not all set.
+     *
+     */
+    template <typename... T0> bool HasNone(T0... aVals) { return (mBits & BuildBits(GetBit(aVals)...)) == 0; }
+
+    /**
+     * This method returns the raw value of the bit flags.
+     *
+     * Note: This is for testing purpose only.
+     *
+     * @returns The raw value of the bit flag.
+     *
+     */
+    BaseType GetRaw(void) { return mBits; }
+
+    /**
+     * This method sets the raw value of the bit flags.
+     *
+     * Note: This is for testing purpose only.
+     *
+     * @param[in] The raw value of the bit flag.
+     *
+     */
+    void SetRaw(BaseType aVal) { mBits = aVal; }
+
+private:
+    // A series of helper functions got actual calculation.
+
+    static BaseType GetBit(EnumType aVal) { return static_cast<BaseType>(1) << static_cast<BaseType>(aVal); };
+
+    template <typename T1, typename... T0> static T1 BuildBits(T1 aVal, T0... aRemaining)
+    {
+        return aVal | BuildBits(aRemaining...);
+    }
+
+    template <typename T1> static T1 BuildBits(T1 aVal) { return aVal; }
+
+    BaseType mBits;
```

Comment:
One topic which we may need to be careful about:
- If we use a simple `uint8_t`/`uint16_t`/`uint32_t` as  bitflags, compilers can optimize it well.
- If we wrap the same `uint` value in a `class`, compilers won't optimize it as effectively  (e.g., won't save/pass it in a register variable and instead pass a pointer to the object). These can slowly add up and increase code size.